src/Core/Lib/WildPatternMatcher.cs

```diff
@@ -45,66 +45,83 @@ public class WildPatternMatcher
         public IEnumerable<int> FindDcp(string T, string P)
         {
             var sa = new SuffixArray<char>(T.ToCharArray());
+            int textLength = T.Length;
             int[] permitted = new int[T.Length];
-            (string[], int[]) pat = SplitPattern(P);
-            var l = pat.Item1.Length;
-            var Val = new int[l];
-            Val[0] = 0;
-            for (int i = 1; i < l; ++i)
-            {
-                Val[i] = Val[i - 1] + pat.Item1[i - 1].Length + pat.Item2[i - 1];
-            }
+            (string[] patternParts, int[] dontCares) = SplitPattern(P);
+            var patternLength = patternParts.Length;
+            var Val = CalculateOffsets(patternParts, dontCares);
             var occ_P = new SortedSet<int>();
-            for (int i = 0; i < l; ++i)
+            for (int i = 0; i < patternLength; ++i)
             {
-                var Pi = pat.Item1[i];
+                var Pi = patternParts[i];
+                var nextPatternWildcards = i<dontCares.Length ? dontCares[i] : 0;
                 var occ_Pi = new SortedSet<int>(sa.FindOccurences(Pi.ToCharArray()));
-                foreach (var r in occ_Pi)
+                foreach (var pos in occ_Pi)
                 {
-                    ++permitted[r - Val[i]];
-                    if (permitted[r - Val[i]] == l)
+                    int adjustedPos = pos - Val[i];
+                    if (!isValidPos(adjustedPos,Val[i], nextPatternWildcards, textLength))
+                        continue;
+                    ++permitted[adjustedPos];
+                    if (permitted[adjustedPos] == patternLength)
                     {
-                        occ_P.Add(r - Val[i]);
+                        occ_P.Add(adjustedPos);
                     }
                 }
             }
             return occ_P;
         }
 
+        private bool isValidPos(int pos,int len,int nextPatternWildcards,int textLength)
```

Comment:
Coding style: for consistency method names are Capitalized. Also, the comment inside the method could be made into a doc comment:
```xml
/// <summary>
/// Checks that a match at the position <paramref name="pos" /> is within the text.
/// </summary>
```